### ISSUE TITLE
Memory leaks due to circular references

### ISSUE THREAD

**Issue by josh-gleason** [jacobgil/pytorch-grad-cam#88]
The current implementation creates circular references that cause major memory leaks. These can be avoided with some minor modifications to the code. To recreate this issue consider the minimal class.

```python
import torch
from torch import nn

class DummyCAM:
    """ dummy CAM class for memory leak demonstration """
    def __init__(self, model, target_layer):
        self.model = model
        self.target_layer = target_layer
        self.target_output = None
        self.target_grad = None
        # registering hooks adds reference to this class into target_layer (circular reference)
        self.target_layer.register_forward_hook(self.save_activation)
        self.target_layer.register_full_backward_hook(self.save_gradient)

    def save_activation(self, module, input, output):
        self.target_output = output

    def save_gradient(self, module, grad_input, grad_output):
        self.target_grad = grad_output[0]

    def __call__(self, input_tensor):
        return self.model(input_tensor)

def foo():
    model = nn.Sequential(
        nn.Linear(2, 3, bias=False),
        nn.Linear(3, 4, bias=False)
    )
    x = torch.ones(1, 2)
    cam = DummyCAM(model, model[1])
    return cam(x)

def main():
    foo()

    # manually execute garbage collector
    import gc
    gc.collect()
    gc.collect()
    # show any dangling tensors
    for obj in gc.get_objects():
        try:
            if torch.is_tensor(obj) or (hasattr(obj, 'data') and torch.is_tensor(obj.data)):
                s = f'{type(obj)}, {obj.size()}'
                print(s)
        except:
            pass

if __name__ == "__main__":
    main()
```

which results in the following tensors being leaked

```none
<class 'torch.Tensor'>, torch.Size([1, 4])
<class 'torch.nn.parameter.Parameter'>, torch.Size([4, 3])
<class 'torch.nn.parameter.Parameter'>, torch.Size([3, 2])
```

Despite the fact that no references to the `model` or `cam` object are available after foo finishes, both of these objects remain in memory.

I'm not actually sure why the Python garbage collector is unable to remove these circular dependencies, but at least on my system this leak is consistent.

That said, I've found a couple of ways to avoid this issue, but at minimum they both require some minor change to the grad-cam code. One way is to store handles for the hooks and then manually unregister them after you are done with the CAM (in the real implementation these would need to be saved at [activations_and_gradients.py#L11](https://github.com/jacobgil/pytorch-grad-cam/blob/f84c8d1bfbe88a5c31b2e5d61d2bf5f8fcaa57a5/pytorch_grad_cam/activations_and_gradients.py#L11)).

```python
class DummyCAM:
    def __init__(self, model, target_layer):
        self.model = model
        self.target_layer = target_layer
        self.target_output = None
        self.target_grad = None
        self.forward_handle = self.target_layer.register_forward_hook(self.save_activation)
        self.backward_handle = self.target_layer.register_full_backward_hook(self.save_gradient)

    def save_activation(self, module, input, output):
        self.target_output = output

    def save_gradient(self, module, grad_input, grad_output):
        self.target_grad = grad_output[0]

    def __call__(self, input_tensor):
        return self.model(input_tensor)

def foo():
    model = nn.Sequential(
        nn.Linear(2, 3, bias=False),
        nn.Linear(3, 4, bias=False)
    )
    x = torch.ones(1, 2)
    cam = DummyCAM(model, model[1])
    result = cam(x)

    # manually unregister (not a great solution but it does work)
    cam.target_output = None
    cam.target_grad = None
    cam.forward_handle.remove()
    cam.backward_handle.remove()

    return result
```

I think a better solution would be to implement some form of context management, implementing `__enter__` and `__exit__` for the `BaseCAM` class to automatically clean up the `ActivationsAndGradients` class upon exit. Using the simple example above this would look something like this

```python
class DummyCAM:
    def __init__(self, model, target_layer):
        self.model = model
        self.target_layer = target_layer
        self.target_output = None
        self.target_grad = None
        self.forward_handle = self.target_layer.register_forward_hook(self.save_activation)
        self.backward_handle = self.target_layer.register_full_backward_hook(self.save_gradient)

    def save_activation(self, module, input, output):
        self.target_output = output

    def save_gradient(self, module, grad_input, grad_output):
        self.target_grad = grad_output[0]

    def __call__(self, input_tensor):
        return self.model(input_tensor)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.target_output = None
        self.target_grad = None
        self.forward_handle.remove()
        self.backward_handle.remove()

def foo():
    model = nn.Sequential(
        nn.Linear(2, 3, bias=False),
        nn.Linear(3, 4, bias=False)
    )
    x = torch.ones(1, 2)
    with DummyCAM(model, model[1]) as cam:
        return cam(x)
```

Such a solution would still allow the previous usage but would also allow usage with a context manager that would automatically clean up after itself.

**Comment by jacobgil** [comment]
Hi,
Thanks a lot for reporting this.

I think the issue is that the PyTorch Backwards hook causes a leakage,
as reported here:
https://github.com/pytorch/pytorch/issues/12863
and documented here in the code:
https://github.com/pytorch/pytorch/blob/25f4b3efe355c9d7d2d9341f818b8616cbb5435b/torch/csrc/autograd/python_variable.cpp#L81-L94

A simple solution to this seems to be adding .detach().
So in the code example you posted above,
self.target_grad = grad_output[0].detach()
Should eliminate the leak.

I pushed an update that does this.

**Comment by josh-gleason** [comment]
Thanks for the quick response and fix. I tested in my larger code-base and this does indeed appear to be a simpler solution. The python garbage collector is now able to detect and remove the unreachable circular reference.

I'll point out that in order to ensure there's no unnecessary memory allocation with your solution I still need to manually call `gc.collect()` after the cam object is no longer reachable. That said, I believe pytorch's memory allocation is robust enough to do this before causing an OOM error (as long as you don't have multiple pytorch processes allocating memory on the same device).
